src/github.com/matrix-org/dendrite/roomserver/input/events.go

```diff
@@ -21,6 +21,21 @@ type RoomEventDatabase interface {
 	// Lookup the Events for a list of numeric event IDs.
 	// Returns a sorted list of events.
 	Events(eventNIDs []int64) ([]types.Event, error)
+	// Lookup the state of a room at each event for a list of string event IDs.
+	// Returns a sorted list of state at each event.
+	// Returns an error if there is an error talking to the database
+	// or if the room state for the event IDs aren't in the database
+	StateAtEventIDs(eventIDs []string) ([]types.StateAtEvent, error)
```

Comment:
Is `len(eventIDs) == len([]types.StateAtEvent)`? If so, I assume we're correlating based on position in the array. This makes me sad.